### DISCUSSION TITLE
Read the remaining time from a timer in RULES

### DISCUSSION THREAD

**Discussion by jkwim** [arendst/Tasmota#19615] [Q&A]
With reference to same requirement that I have, I found the following old post:

 [RuleTimers - timeleft #3571 ]https://github.com/arendst/Tasmota/issues/3571(url)
 
 However it is not clear to me how the event TIMELEFT would be genetated with a value for a given timer eg: T3
 
 What I really need is to check for remaining time periodically and fire another action.
 
 Say I set timer T3 to 600 seconds.
 
 I want to catch the events at say `T3=540, T3=480, T3=420 `etc.
 
 One way based on above post is to use `SetOption4 1 `and then catch 
`ON ruletimer#T3=540 DO ... ENDON
`
However this might work only if `ruletimer `command has been manually run right?

I need a method which continuously evalues the value of T3 and then trigger action when the value reaches certain values.

Any help?

**Comment by sfromis** [comment]
When you use rules to issue commands to retrieve values from the result JSON, you need to "wrap" it in Backlog for the triggers to work. As the output of "Backlog RuleTimer" does not include the key "RuleTimer" as part of the payload, the pertinent trigger format to grab a value would be something like `on T3==540 do ... endon`. However, this is a bad way, as it will only trigger in case of the seconds left being exactly 540. "Less bad" would be T3<540 (obviously depending on circumstances) but still not a good strategy for timers. A better way would be to have an additional ruletimer expiring at the pertinent time.

**Reply by jkwim** [reply]
Thanks for that quick guidence. I understood why expecting an exact value match might not happen.

In my case, the T3 trigger is coming in as a MQTT command message.

And I see the following RULETIMER trigger response:

`stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":899,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}`

Can I trap the above event as just `T3!=0` to trigger a second timer?

`ON T3!=0 DO ruletimer4 60 ENDON`

**Reply by sfromis** [reply]
Triggering on != is technically workable. but risks triggering multiple times. What I was suggesting was to issue both RuleTimer commands at the same time, instead of complicating things.

Not sure what you mean by "coming in as a MQTT command message". Tasmota rules are not acting on incoming MQTT messages, but on **outgoing** ones with JSON, including the message you get when you issue a RuleTimer command. To act on incoming messages, simplest is to have the message be an `Event` command with an associated trigger. Tasmota does also have a feature for subscribing to messages, and picking up single JSON value for a rule trigger, if you use a build with this feature included.

What I wrote about with T3 was based on this appearing in the output of the RuleTimer command, and if this is not what you mean, I'd recommend using another name to avoid confusion.

**Reply by jkwim** [reply]
I was just testing and it seems to work.

I am sending a command like `cmnd/tasmota/RULETIMER3 900` via MQTT.

On the target device I see the following:

`stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":899,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}`

and I get the trigger

`RUL: T3!=0 performs "ruletimer4 30"`
 
I suppose I can build my logic from there

**Reply by jkwim** [reply]
Update:
I see your point about multiple triggers from other instances of RULETIMER messages

In that case a condition like `T3>0` should prevent it right?

**Reply by jkwim** [reply]
Update2:

Ah! it does not solve the issue. I suppose that I need to go the event way then.

Let me try

**Reply by sfromis** [reply]
In general, no, but possibly in some particular situation. Still a very "fragile" technique easy to mess up next time you change something, and `>0` is likely not different from `!=0`. The thing is that if you set another RuleTimer, all triggers on the whole list of 8 timers are evaluated again. Hence I do not see rules reacting to those messages as a sound technique. IMO.

**Comment by jkwim** [comment]
Finally got this working:

Objective:
Once the `DELAY` is set I need to get a periodic notification of remaning number of intervals. (at the moment the last remaning interval can be shorter than the interval but that is ok for my requirement)

 

`var2` stores the interval between the notification loop
`var3` just picks the value of the delay but not used in the rule yet
`var4` stores the calulated number of remaining intervals of duration `var2`

```
ON rules#timer=3 DO Backlog POWER %var1%; var4 0; ruletimer4 0 ENDON
ON rules#timer=4 DO Backlog sub4 1; ruletimer4 %var2%; ENDON
ON EVENT#DELAY DO Backlog var2 60; ruletimer3 %value%; ruletimer4 %var2%; var3 %value%; CalcRes 0; scale4 %var3%,0,%var2%,0,1 ENDON
```
`EVENT DELAY=100 `can be fired manually or via MQTT

Here is the output:

```
02:33:13.571 CMD: EVENT DELAY=100
02:33:13.578 MQT: stat/tasmota/EVENT = {"Event":"Done"}
02:33:13.649 RUL: EVENT#DELAY performs "Backlog var2 60; ruletimer3 100; ruletimer4 60; var3 100; CalcRes 0; scale4 100,0,60,0,1"
02:33:13.675 MQT: stat/tasmota/VAR = {"Var2":"60"}
02:33:13.881 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":99,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
02:33:14.095 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":99,"T4":60,"T5":0,"T6":0,"T7":0,"T8":0}
02:33:14.344 MQT: stat/tasmota/VAR = {"Var3":"100"}
02:33:14.551 MQT: stat/tasmota/CALCRES = {"CalcRes":0}
02:33:14.802 MQT: stat/tasmota/SCALE = {"Scale4":"2"}

<<<<< after 60secs

02:34:14.296 RUL: RULES#TIMER=4 performs "Backlog sub4 1; ruletimer4 60;"
02:34:14.330 MQT: stat/tasmota/SUB = {"Sub4":"1"}
02:34:14.538 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":39,"T4":59,"T5":0,"T6":0,"T7":0,"T8":0}

<<<<< after 40secs

02:34:54.315 RUL: RULES#TIMER=3 performs "Backlog POWER 1; var4 0; ruletimer4 0"
02:34:54.396 MQT: stat/tasmota/POWER = {"POWER":"ON"}
02:34:54.399 MQT: stat/tasmota/POWER = ON
02:34:54.570 MQT: stat/tasmota/VAR = {"Var4":"0"}
02:34:54.779 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
```

This does what I needed for the time being.

Thanks a lot for your quick help!

**Reply by sfromis** [reply]
Beware that first setting and then using %var2% in the same delay rule means that it is using the old value of var2 instead of the new. As CalcRes is a persistent setting, no benefit from setting it repeatedly.

**Reply by jkwim** [reply]
Yes, while attempting to improve/evolve the script I figured that out.

I am thinking about using VARx#state to trigger the further actions.

Still working on it

**Comment by jkwim** [comment]
Well this seems to work.

Over to you for your valuable comments.

The buzzer is physically a remote device.

`var2` & `var4` will be initialized at boot time.
So will be `CalcRes 0`.
`var1` is handled by another rule.

```
ON rules#timer=3 DO Backlog POWER %var1%; var4 0; ruletimer4 0; PUBLISH cmnd/tasmota2/BUZZER "1,10,1" ENDON
ON rules#timer=4 DO Backlog sub4 1; ruletimer4 %var2% ENDON
ON var4#STATE>0 DO PUBLISH cmnd/tasmota2/BUZZER %var4% ENDON
ON EVENT#DELAY DO Backlog ruletimer3 %value%; ruletimer4 %var2%; var3 %value%; scale4 %var3%,0,%var2%,0,1 ENDON
```


```
11:27:56.664 CMD: EVENT DELAY=100
11:27:56.672 MQT: stat/tasmota/EVENT = {"Event":"Done"}
11:27:56.743 RUL: EVENT#DELAY performs "Backlog ruletimer3 100; ruletimer4 60; var3 100; scale4 100,0,60,0,1"
11:27:56.772 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":100,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
11:27:57.024 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":99,"T4":60,"T5":0,"T6":0,"T7":0,"T8":0}
11:27:57.223 MQT: stat/tasmota/VAR = {"Var3":"100"}
11:27:57.434 MQT: stat/tasmota/SCALE = {"Scale4":"2"}
11:27:57.476 RUL: VAR4#STATE>0 performs "PUBLISH cmnd/tasmota2/BUZZER 2"
11:27:57.480 MQT: cmnd/tasmota2/BUZZER = 2

11:28:57.770 RUL: RULES#TIMER=4 performs "Backlog sub4 1; ruletimer4 60"
11:28:57.804 MQT: stat/tasmota/SUB = {"Sub4":"1"}
11:28:57.848 RUL: VAR4#STATE>0 performs "PUBLISH cmnd/tasmota2/BUZZER 1"
11:28:57.851 MQT: cmnd/tasmota2/BUZZER = 1
11:28:58.079 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":38,"T4":60,"T5":0,"T6":0,"T7":0,"T8":0}

11:29:37.776 RUL: RULES#TIMER=3 performs "Backlog POWER 0; var4 0; ruletimer4 0; PUBLISH cmnd/tasmota2/BUZZER "1,10,1""
11:29:37.853 MQT: stat/tasmota/POWER = {"POWER":"OFF"}
11:29:37.856 MQT: stat/tasmota/POWER = OFF
11:29:38.031 MQT: stat/tasmota/VAR = {"Var4":"0"}
11:29:38.236 MQT: stat/tasmota/RULETIMER = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
11:29:38.531 MQT: cmnd/tasmota2/BUZZER = "1,10,1"
```

What it does is that when a delayed start is initiated, periodically the buzzer will indicate the remaining number of intervals. Ofcourse the last interval might be shorter before the final action occurs (I might add another timer to indicate that with a shorter interval beeps).

**Reply by sfromis** [reply]
Using `"` here is a syntax error: `BUZZER "1,10,1"`
And you are still using `var3` in the Scale command after setting it in the same rule.

**Comment by jkwim** [comment]
Based on your feedback, I tried to build a chain of EVENTs to handle the variable assignments and calculations.

Here is the outcome:

```
Rule2
ON EVENT#DELAY DO Backlog ruletimer3 %value%;  var3 %value%; EVENT SETVAR2=1 ENDON
ON EVENT#SETVAR2 DO Backlog var2 60; EVENT SETTIMER4=1 ENDON
ON EVENT#SETTIMER4 DO ruletimer4 %var2% ENDON
ON var2#STATE DO scale5 %var2%,0,10,0,1 ENDON
ON var3#STATE DO scale4 %var3%,0,%var2%,0,1 ENDON
ON var4#STATE>0 DO PUBLISH cmnd/tasmota2/BUZZER %var4% ENDON
ON var4#STATE==1 DO ruletimer5 %var5% ENDON
ON rules#timer=3 DO Backlog POWER %var1%; var4 0; var5 0; ruletimer4 0; ruletimer5 0; PUBLISH cmnd/tasmota2/BUZZER "1,10,1" ENDON
ON rules#timer=4 DO Backlog sub4 1; ruletimer4 %var2% ENDON
ON rules#timer=5 DO Backlog ruletimer5 %var5%; PUBLISH cmnd/tasmota2/BUZZER "1,4,2" ENDON
```

```
18:34:41.072 CMD: EVENT DELAY=100
18:34:41.079 MQT: stat/tasmota/EVENT = {"Event":"Done"}
18:34:41.154 RUL: EVENT#DELAY performs "Backlog ruletimer3 100;  var3 100; EVENT SETVAR2=1"
18:34:41.191 MQT: stat/tasmota/RULETIMER = {"T1":41,"T2":0,"T3":100,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
18:34:41.442 MQT: stat/tasmota/VAR = {"Var3":"100"}
18:34:41.491 RUL: VAR3#STATE performs "scale4 100,0,60,0,1"
18:34:41.499 MQT: stat/tasmota/SCALE = {"Scale4":"2"}
18:34:41.549 RUL: VAR4#STATE>0 performs "PUBLISH cmnd/tasmota2/BUZZER 2"
18:34:41.557 MQT: cmnd/tasmota2/BUZZER = 2
18:34:41.789 MQT: stat/tasmota/EVENT = {"Event":"Done"}
18:34:41.836 RUL: EVENT#SETVAR2 performs "Backlog var2 60; EVENT SETTIMER4=1"
18:34:41.871 MQT: stat/tasmota/VAR = {"Var2":"60"}
18:34:41.918 RUL: VAR2#STATE performs "scale5 60,0,10,0,1"
18:34:41.924 MQT: stat/tasmota/SCALE = {"Scale5":"6"}
18:34:42.155 MQT: stat/tasmota/EVENT = {"Event":"Done"}
18:34:42.199 RUL: EVENT#SETTIMER4 performs "ruletimer4 60"
18:34:42.205 MQT: stat/tasmota/RULETIMER = {"T1":40,"T2":0,"T3":98,"T4":59,"T5":0,"T6":0,"T7":0,"T8":0}

<=============== AFTER 60 SECONDS ==================>

18:35:42.354 RUL: RULES#TIMER=4 performs "Backlog sub4 1; ruletimer4 60"
18:35:42.379 MQT: stat/tasmota/SUB = {"Sub4":"1"}
18:35:42.430 RUL: VAR4#STATE>0 performs "PUBLISH cmnd/tasmota2/BUZZER 1"
18:35:42.437 MQT: cmnd/tasmota2/BUZZER = 1


18:35:42.444 RUL: VAR4#STATE==1 performs "ruletimer5 6"
18:35:42.450 MQT: stat/tasmota/RULETIMER = {"T1":281,"T2":0,"T3":38,"T4":0,"T5":6,"T6":0,"T7":0,"T8":0}
18:35:42.686 MQT: stat/tasmota/RULETIMER = {"T1":281,"T2":0,"T3":38,"T4":60,"T5":5,"T6":0,"T7":0,"T8":0}

<=============== AFTER 6 SECONDS ==================>

18:35:49.350 RUL: RULES#TIMER=5 performs "Backlog ruletimer5 6; PUBLISH cmnd/tasmota2/BUZZER "1,4,2""
18:35:49.376 MQT: stat/tasmota/RULETIMER = {"T1":274,"T2":0,"T3":31,"T4":53,"T5":5,"T6":0,"T7":0,"T8":0}
18:35:49.579 MQT: cmnd/tasmota2/BUZZER = "1,4,2"

<=============== AFTER 6 SECONDS ==================>

18:35:56.357 RUL: RULES#TIMER=5 performs "Backlog ruletimer5 6; PUBLISH cmnd/tasmota2/BUZZER "1,4,2""
18:35:56.380 MQT: stat/tasmota/RULETIMER = {"T1":267,"T2":0,"T3":24,"T4":46,"T5":6,"T6":0,"T7":0,"T8":0}
18:35:56.623 MQT: cmnd/tasmota2/BUZZER = "1,4,2"

<=============== AFTER 6 SECONDS ==================>

18:36:03.363 RUL: RULES#TIMER=5 performs "Backlog ruletimer5 6; PUBLISH cmnd/tasmota2/BUZZER "1,4,2""
18:36:03.386 MQT: stat/tasmota/RULETIMER = {"T1":260,"T2":0,"T3":17,"T4":39,"T5":6,"T6":0,"T7":0,"T8":0}
18:36:03.583 MQT: cmnd/tasmota2/BUZZER = "1,4,2"

<=============== AFTER 6 SECONDS ==================>

18:36:10.349 RUL: RULES#TIMER=5 performs "Backlog ruletimer5 6; PUBLISH cmnd/tasmota2/BUZZER "1,4,2""
18:36:10.373 MQT: stat/tasmota/RULETIMER = {"T1":253,"T2":0,"T3":10,"T4":32,"T5":5,"T6":0,"T7":0,"T8":0}
18:36:10.573 MQT: cmnd/tasmota2/BUZZER = "1,4,2"

<=============== AFTER 6 SECONDS ==================>

18:36:17.336 RUL: RULES#TIMER=5 performs "Backlog ruletimer5 6; PUBLISH cmnd/tasmota2/BUZZER "1,4,2""
18:36:17.361 MQT: stat/tasmota/RULETIMER = {"T1":246,"T2":0,"T3":3,"T4":25,"T5":5,"T6":0,"T7":0,"T8":0}
18:36:17.561 MQT: cmnd/tasmota2/BUZZER = "1,4,2"

<=============== AFTER 4 SECONDS FINAL BUZZER ==================>

18:36:21.364 RUL: RULES#TIMER=3 performs "Backlog POWER 1; var4 0; var5 0; ruletimer4 0; ruletimer5 0; PUBLISH cmnd/tasmota2/BUZZER "1,10,1""
18:36:21.436 MQT: stat/tasmota/POWER = {"POWER":"ON"}
18:36:21.439 MQT: stat/tasmota/POWER = ON
18:36:21.632 MQT: stat/tasmota/VAR = {"Var4":"0"}
18:36:21.847 MQT: stat/tasmota/VAR = {"Var5":"0"}
18:36:22.064 MQT: stat/tasmota/RULETIMER = {"T1":241,"T2":0,"T3":0,"T4":0,"T5":1,"T6":0,"T7":0,"T8":0}
18:36:22.266 MQT: stat/tasmota/RULETIMER = {"T1":241,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
18:36:22.520 MQT: cmnd/tasmota2/BUZZER = "1,10,1"
```

About the syntax violation of `PUBLISH cmnd/tasmota2/BUZZER "1,4,2"`
It does work.
The reason for using quotes was that I did not see the command being recived by the other device when I did not use quotes. I will test this again.
